Do not rely on expdef_**** files (e.g. in autosubmit_git for oper experiments)

### Description

`expdef_***.yml` files were always present in AS3 experiments. In AS4 experiments users may use whatever they want, and in the end we parse & combine them, producing the `conf/metadata/experiment_data.yml` file, that shows what AS has in memory.

See this comment for an example added in 4.1.14: https://github.com/BSC-ES/autosubmit/pull/2293/files#r2182051227

* Autosubmit version: 4.1.14, 4.1.15
* Machine/VM/environment name (if applicable): ClimateDT
* Experiment ID (if applicable):o00b

### Reproducible Example

<!--
  Please provide an example of how to replicate this bug to help us diagnose
  the issue.
-->

Create an `o---` experiment, using Git, and minimal. Try running it, and you should get an exception similar to:

```bash
2025-07-03 09:49:17,676 Autosubmit is running with 4.1.15
[WARNING] 2025-07-03 09:49:17,781 Yaml file /appl/AS/AUTOSUBMIT_DATA/o00b/proj/git_project/conf/additional_jobs/transfer-False.yml not found
[ERROR] 2025-07-03 09:49:18,276 Arguments provided: Namespace(version=False, logfile='DEBUG', logconsole='WARNING', command='run', expid='o00b', notransitive=False, update_version=False, start_time=None, start_after=None, run_only_members=None, profile=False)
[ERROR] 2025-07-03 09:49:18,277 This is the experiment: <o00b> which had an issue with the command: <run> and it is currently using the Autosubmit Version: 4.1.15.
2025-07-03 09:49:18,277 Traceback (most recent call last):
  File "/appl/AS/4.1.15/lib64/python3.9/site-packages/autosubmit/scripts/autosubmit.py", line 105, in main
    return_value = Autosubmit.run_command(args)
  File "/appl/AS/4.1.15/lib64/python3.9/site-packages/autosubmit/autosubmit.py", line 724, in run_command
    AutosubmitGit.check_unpushed_changes(expid)
  File "/appl/AS/4.1.15/lib64/python3.9/site-packages/autosubmit/git/autosubmit_git.py", line 290, in check_unpushed_changes
    with open(origin, 'r') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/appl/AS/AUTOSUBMIT_DATA/o00b/conf/expdef_o00b.yml'

[CRITICAL] 2025-07-03 09:49:18,277 Unexpected error: [Errno 2] No such file or directory: '/appl/AS/AUTOSUBMIT_DATA/o00b/conf/expdef_o00b.yml'.
 Please report it to Autosubmit Developers through Git
2025-07-03 09:49:18,277 More info at https://autosubmit.readthedocs.io/en/master/troubleshooting/error-codes.html
```

### Expected Behaviour

<!-- Please describe what you expected to happen. -->
No exception raised, and instead a nice message displayed to user if appropriate.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,11 @@
 ### 4.1.16: Unreleased
 
+**Bug fixes:**
+
+- Fixed issue with the verification of dirty Git local repositories in operational experiments #2446
+
+**Enhancements:**
+
 - autosubmit/autosubmit container now includes the `$USER` environment variable
   via its entrypoint #2359
 - Adding a Slurm Container to the CI/CD and creating tests to increase the

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -141,3 +141,12 @@ or just the tests that require Slurm:
 ```bash
 $ pytest -m 'slurm'
 ```
+
+## Random ports
+
+Some tests require random ports. This table below can be useful for troubleshooting
+if we ever start running out of ports.
+
+- The integration test `test_paramiko_platform.py` uses the range `2500` to `3000`
+- The integration test `test_mail.py` uses the range `3500` to `4000`
+- The integration test `test_autosubmit_git.py` uses the range `4000` to `4500`

--- a/autosubmit/autosubmit.py
+++ b/autosubmit/autosubmit.py
@@ -50,7 +50,6 @@ from pyparsing import nestedExpr
 from ruamel.yaml import YAML
 
 import autosubmit.helpers.autosubmit_helper as AutosubmitHelper
-import autosubmit.history.utils as HUtils
 import autosubmit.statistics.utils as StatisticsUtils
 from autosubmit.database.db_common import create_db
 from autosubmit.database.db_common import delete_experiment, get_experiment_descrip
@@ -60,7 +59,7 @@ from autosubmit.database.db_structure import get_structure
 from autosubmit.experiment.detail_updater import ExperimentDetails
 from autosubmit.experiment.experiment_common import copy_experiment
 from autosubmit.experiment.experiment_common import new_experiment
-from autosubmit.git.autosubmit_git import AutosubmitGit
+from autosubmit.git.autosubmit_git import AutosubmitGit, check_unpushed_changes, clean_git
 from autosubmit.helpers.processes import process_id
 from autosubmit.helpers.utils import check_jobs_file_exists, get_rc_path
 from autosubmit.helpers.utils import strtobool
@@ -719,9 +718,7 @@ class Autosubmit:
             expid = args.expid
         if args.command != "configure" and args.command != "install":
             Autosubmit._init_logs(args, args.logconsole, args.logfile, expid)
-
         if args.command == 'run':
-            AutosubmitGit.check_unpushed_changes(expid)
             return Autosubmit.run_experiment(args.expid, args.notransitive,args.start_time,args.start_after, args.run_only_members, args.profile)
         elif args.command == 'expid':
             return Autosubmit.expid(args.description,args.HPC,args.copy, args.dummy,args.minimal_configuration,args.git_repo,args.git_branch,args.git_as_conf,args.operational,args.testcase,args.evaluation,args.use_local_minimal) != ''
@@ -2240,6 +2237,16 @@ class Autosubmit:
                     raise
                 except Exception as e:
                     raise AutosubmitCritical("Error in run initialization", 7014, str(e))  # Changing default to 7014
+
+                as_conf_config = as_conf.experiment_data.get('CONFIG', {})
+                git_operational_check_enabled = as_conf_config.get('GIT_OPERATIONAL_CHECK_ENABLED', True)
+
+                if git_operational_check_enabled:
+                    Log.debug('Checking for dirty local Git repository')
+                    check_unpushed_changes(expid, as_conf)
+                else:
+                    Log.warning('Git operational check disabled by user')
+
                 Log.debug("Running main running loop")
                 did_run = False
                 #########################
@@ -2952,9 +2959,8 @@ class Autosubmit:
                 if project_type == "git":
                     Log.info("Registering commit SHA...")
                     autosubmit_config.set_git_project_commit(autosubmit_config)
-                    autosubmit_git = AutosubmitGit(expid)
                     Log.info("Cleaning GIT directory...")
-                    if not autosubmit_git.clean_git(autosubmit_config):
+                    if not clean_git(autosubmit_config):
                         return False
                     Log.result("Git project cleaned!\n")
                 else:

--- a/autosubmit/git/autosubmit_git.py
+++ b/autosubmit/git/autosubmit_git.py
@@ -1,38 +1,33 @@
-#!/usr/bin/env python3
-
-# Copyright 2015-2020 Earth Sciences Department, BSC-CNS
-
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
 # This file is part of Autosubmit.
-
+#
 # Autosubmit is free software: you can redistribute it and/or modify
 # it under the terms of the GNU General Public License as published by
 # the Free Software Foundation, either version 3 of the License, or
 # (at your option) any later version.
-
+#
 # Autosubmit is distributed in the hope that it will be useful,
 # but WITHOUT ANY WARRANTY; without even the implied warranty of
 # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 # GNU General Public License for more details.
-
-from os import path
-
+#
 # You should have received a copy of the GNU General Public License
 # along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
 import locale
 import os
-import psutil
 import re
 import shutil
 import subprocess
 from pathlib import Path
-from ruamel.yaml import YAML
 from shutil import rmtree
 from time import time
-from typing import List, Union
+from typing import List, Optional, Union
 
-# from autosubmit import Autosubmit
-from autosubmit.helpers.processes import process_id
 from autosubmitconfigparser.config.basicconfig import BasicConfig
+from autosubmitconfigparser.config.configcommon import AutosubmitConfig
+
 from log.log import Log, AutosubmitCritical
 
 Log.get_logger("Autosubmit")
@@ -45,57 +40,122 @@ _GIT_URL_PATTERN = re.compile(
     )''', re.VERBOSE)
 """Regular expression to match Git URL."""
 
+_GIT_UNCOMMITTED_CMD = ('git', 'status', '--porcelain')
+"""Command to check if there are changes not committed go local Git repository."""
+
+_GIT_UNPUSHED_CMD = ('git', 'log', '--branches', '--not', '--remotes')
+"""Command to check if there are changes not pushed to Git remotes."""
+
+
+def _get_uncommitted_code(git_repo: Path) -> Optional[str]:
+    """Return any uncommitted changes in the given Git repository or submodules.
+
+    If there are no uncommitted changes, return None.
+
+    :param git_repo: Path to the Git repository.
+    :return: A string containing the list of code not committed (directly git command output).
+    Returns``None`` if there are no uncommitted changes.
+    """
+    encoding = locale.getlocale()[1]
+
+    git_output = subprocess.check_output(_GIT_UNCOMMITTED_CMD, cwd=git_repo, encoding=encoding)
+    for line in git_output.splitlines():
+        tokens = line.strip().split(' ')
+        if tokens[0] != '':
+            return git_output
+
+    return None
+
+
+def _get_code_not_pushed(git_repo: Path) -> Optional[str]:
+    """Return any code not pushed to remotes in the given Git repository or submodules.
+
+    If there are no code not pushed changes, return None.
+
+    :param git_repo: Path to the Git repository.
+    :return: A string containing the list of code not pushed (directly git command output).
+    Returns``None`` if there are no pushed not changes.
+    """
+    encoding = locale.getlocale()[1]
+
+    git_output = subprocess.check_output(_GIT_UNPUSHED_CMD, cwd=git_repo, encoding=encoding)
+    for line in git_output.splitlines():
+        if line.strip():
+            return git_output
+
+    return None
+
+
+def check_unpushed_changes(expid: str, as_conf: AutosubmitConfig) -> None:
+    """Check if the Git repository is dirty for an operational experiment.
+
+    Raises an AutosubmitCritical error if the experiment is operational,
+    the platform is Git, and there are unpushed changes in the local Git
+    repository, or in any of its Git submodules.
+
+    :param expid: The experiment ID.
+    :param as_conf: Autosubmit configuration object.
+    """
+    project_type = as_conf.get_project_type()
+    if expid[0] == 'o' and project_type == 'git':
+        proj_dir = Path(as_conf.get_project_dir())
+
+        if uncommitted_code := _get_uncommitted_code(proj_dir):
+            message = ("You must commit and push your code to the remote Git repository in an "
+                       f"operational experiment before running it.\n\n{uncommitted_code}")
+            raise AutosubmitCritical(message, 7075)
+
+        if not_pushed := _get_code_not_pushed(proj_dir):
+            message = ("You must push your code to the remote Git repository in an "
+                       f"operational experiment before running it.\n\n{not_pushed}")
+            raise AutosubmitCritical(message, 7075)
+
+
+def clean_git(as_conf: AutosubmitConfig) -> bool:
+    """Clean the cloned Git repository inside the project directory of the experiment.
+
+    Skipped if the project directory location is not a valid directory.
+
+    Skipped if the project directory is not a valid Git repository.
+
+    Skipped if there are changes in the Git repository that were not committed or
+    not pushed.
+
+    :param as_conf: experiment configuration
+    :return: ``True`` if the Git project directory was successfully deleted, ``False`` otherwise.
+    """
+    dirname_path = Path(as_conf.get_project_dir())
+    Log.debug("Checking git directory status...")
+
+    if not dirname_path.is_dir():
+        Log.debug("Not a directory... SKIPPING!")
+        return False
+
+    if not Path(dirname_path, '.git').is_dir():
+        Log.debug("Not a git repository... SKIPPING!")
+        return False
+
+    if _get_uncommitted_code(dirname_path):
+        Log.info("Changes not committed detected... SKIPPING!")
+        raise AutosubmitCritical("Commit needed!", 7013)
+
+    if _get_code_not_pushed(dirname_path):
+        Log.info("Changes not pushed detected... SKIPPING!")
+        raise AutosubmitCritical("Synchronization needed!", 7064)
+
+    if not as_conf.set_git_project_commit(as_conf):
+        Log.info("Failed to set Git project commit... SKIPPING!")
+        return False
+
+    proj_dir = Path(BasicConfig.LOCAL_ROOT_DIR, as_conf.expid, BasicConfig.LOCAL_PROJ_DIR)
+    Log.debug(f"Removing project directory {str(proj_dir)}")
+    rmtree(proj_dir)
+
+    return True
+
+
 class AutosubmitGit:
-    """
-    Class to handle experiment git repository
-
-    :param expid: experiment identifier
-    :type expid: str
-    """
-
-    def __init__(self, expid):
-        self._expid = expid
-
-    @staticmethod
-    def clean_git(as_conf):
-        """
-        Function to clean space on BasicConfig.LOCAL_ROOT_DIR/git directory.
-
-        :param as_conf: experiment configuration
-        :type as_conf: autosubmitconfigparser.config.AutosubmitConfig
-        """
-        proj_dir = os.path.join(
-            BasicConfig.LOCAL_ROOT_DIR, as_conf.expid, BasicConfig.LOCAL_PROJ_DIR)
-        dirname_path = as_conf.get_project_dir()
-        Log.debug("Checking git directory status...")
-        if path.isdir(dirname_path):
-            if path.isdir(os.path.join(dirname_path, '.git')):
-                try:
-                    output = subprocess.check_output("cd {0}; git diff-index HEAD --".format(dirname_path),
-                                                     shell=True)
-                except subprocess.CalledProcessError as e:
-                    raise AutosubmitCritical(
-                        "Failed to retrieve git info ...", 7064, str(e))
-                if output:
-                    Log.info("Changes not committed detected... SKIPPING!")
-                    raise AutosubmitCritical("Commit needed!", 7013)
-                else:
-                    output = subprocess.check_output("cd {0}; git log --branches --not --remotes".format(dirname_path),
-                                                     shell=True)
-                    if output:
-                        Log.info("Changes not pushed detected... SKIPPING!")
-                        raise AutosubmitCritical(
-                            "Synchronization needed!", 7064)
-                    else:
-                        if not as_conf.set_git_project_commit(as_conf):
-                            return False
-                        Log.debug("Removing directory")
-                        rmtree(proj_dir)
-            else:
-                Log.debug("Not a git repository... SKIPPING!")
-        else:
-            Log.debug("Not a directory... SKIPPING!")
-        return True
+    """Class to handle experiment git repository."""
 
     @staticmethod
     def clone_repository(as_conf, force, hpcarch):
@@ -259,9 +319,11 @@ class AutosubmitGit:
             if os.path.exists(project_backup_path):
                 Log.info("Restoring proj folder...")
                 shutil.move(project_backup_path, project_path)
-            raise AutosubmitCritical(f'Can not clone {git_project_branch+" "+git_project_origin} into {project_path}', 7065)
+            raise AutosubmitCritical(
+                f'Can not clone {git_project_branch + " " + git_project_origin} into {project_path}', 7065)
         if submodule_failure:
-            Log.info("Some Submodule failures have been detected. Backup {0} will not be removed.".format(project_backup_path))
+            Log.info("Some Submodule failures have been detected. Backup {0} will not be removed.".format(
+                project_backup_path))
             return False
 
         if os.path.exists(project_backup_path):
@@ -275,45 +337,3 @@ class AutosubmitGit:
         git_repo = git_repo.lower().strip()
 
         return _GIT_URL_PATTERN.match(git_repo) is not None
-
-    @staticmethod
-    def check_unpushed_changes(expid: str) -> None:
-        """
-        Raises an AutosubmitCritical error if the experiment is operational, the platform is Git, and there are unpushed changes.
-
-        Args: expid (str): The experiment ID.
-
-        Returns: None
-        """
-        if expid[0] == 'o':
-            origin = Path(BasicConfig.expid_dir(expid).joinpath("conf/expdef_{}.yml".format(expid)))
-            with open(origin, 'r') as f:
-                yaml = YAML(typ='rt')
-                data = yaml.load(f)
-                project = data["PROJECT"]["PROJECT_TYPE"]
-
-            version_controls = ["git", 
-                                "git submodule"]
-            arguments = [["status", "--porcelain"],
-                        ["foreach", "'git status --porcelain'"]]
-            for version_control, args in zip(version_controls, arguments):
-                if project == version_control:
-                    output = subprocess.check_output([version_control, args]).decode(locale.getlocale()[1])
-                    if any(status.startswith(code) for code in ["M", "A", "D", "?"] for status in output.splitlines()):
-                        # M: Modified, A: Added, D: Deleted, ?: Untracked
-                        raise AutosubmitCritical("Push local changes to remote repository before running", 7075)
-
-    @staticmethod
-    def check_directory_in_use(expid: str) -> bool:
-        """
-        Checks for open files and unfinished git-credential requests in a directory
-        """
-        if process_id(expid) is not None:
-            return True
-        for proc in psutil.process_iter(['name', 'cmdline']):
-            name = proc.info.get('name', '')
-            cmdline = proc.info.get('cmdline', [])
-            if '_run.log' in name or any('run' in arg for arg in cmdline): 
-                return True
-        return False 
-

--- a/docs/source/userguide/run/index.rst
+++ b/docs/source/userguide/run/index.rst
@@ -23,12 +23,6 @@ Options:
 
 Example:
 
-.. code-block:: bash
-
-    # Add your key to ssh agent ( if encrypted )
-    ssh-add ~/.ssh/id_rsa
-    autosubmit run <EXPID>
-
 .. important:: If the autosubmit version is set on ``autosubmit_<EXPID>.yml`` it must match the actual autosubmit version
 .. hint:: It is recommended to launch it in background and with ``nohup`` (continue running although the user who launched the process logs out).
 
@@ -48,6 +42,17 @@ Example:
 More info on password-less ssh can be found at: http://www.linuxproblem.org/art_9.html
 
 .. caution:: After launching Autosubmit, one must be aware of login expiry limit and policy (if applicable for any HPC) and renew the login access accordingly (by using token/key etc) before expiry.
+
+When running operational experiments (i.e. an experiment whose EXPID starts with ``'o'``, e.g. ``o001``),
+and that have a Git project, Autosubmit checks if there is any code that was not committed
+or not pushed to the remote Git repository.
+
+If there are local changes not committed and pushed, Autosubmit will fail to run
+the experiment, print an error message, and exit with an exit code different than zero.
+
+This can be disabled by setting the property ``CONFIG.GIT_OPERATIONAL_CHECK_ENABLED``
+to ``False`` (it is ``True`` by default). Note, however, that this is discouraged as
+it would affect the traceability of operational experiments.
 
 How to run an experiment that was created with another version
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

--- a/test/conftest.py
+++ b/test/conftest.py
@@ -15,24 +15,27 @@
 # You should have received a copy of the GNU General Public License
 # along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
 
-# Fixtures available to multiple test files must be created in this file.
+"""Fixtures available to multiple test files must be created in this file."""
+
 import os
 import pwd
 from dataclasses import dataclass
 from datetime import datetime
 from fileinput import FileInput
 from pathlib import Path
+from random import randrange
 from random import seed, randint, choice
 from re import sub
 from textwrap import dedent
 from time import time
-from typing import TYPE_CHECKING, Any, Dict, Protocol, Optional, Type, List
+from typing import TYPE_CHECKING, Any, Callable, Dict, Generator, List, Optional, Protocol, Tuple, Type
 
 import pytest
 from autosubmitconfigparser.config.basicconfig import BasicConfig
 from autosubmitconfigparser.config.configcommon import AutosubmitConfig
 from pytest_mock import MockerFixture
 from ruamel.yaml import YAML
+from testcontainers.sftp import DockerContainer, wait_for_logs
 
 from autosubmit.autosubmit import Autosubmit
 from autosubmit.job.job import Job
@@ -101,6 +104,7 @@ def _initialize_autosubmitrc(folder: Path) -> Path:
 
 class AutosubmitExperimentFixture(Protocol):
     """Type for ``autosubmit_exp`` fixture."""
+
     def __call__(
             self,
             expid: Optional[str] = None,
@@ -139,7 +143,7 @@ def autosubmit_exp(
     def _create_autosubmit_exp(
             expid: Optional[str] = None,
             experiment_data: Optional[Dict] = None,
-            wrapper = False,
+            wrapper=False,
             *_,
             **kwargs
     ):
@@ -591,3 +595,30 @@ def create_jobs(
         return jobs
 
     return _create_jobs(mocker, request.param[0], request.param[1])
+
+
+@pytest.fixture(scope="function")
+def git_server(tmp_path) -> Generator[Tuple[DockerContainer, Path, str], None, None]:
+
+    # Start a container to server it -- otherwise, we would have to use
+    # `git -c protocol.file.allow=always submodule ...`, and we cannot
+    # change how Autosubmit uses it in `autosubmit create` (due to bad
+    # code design choices).
+
+    git_repos_path = tmp_path / 'git_repos'
+    git_repos_path.mkdir(exist_ok=True, parents=True)
+
+    http_port = randrange(4000, 4500)
+
+    image = 'githttpd/githttpd:latest'
+    with DockerContainer(image=image, remove=True) \
+            .with_bind_ports(80, http_port) \
+            .with_volume_mapping(str(git_repos_path), '/opt/git-server', mode='rw') as container:
+        wait_for_logs(container, "Command line: 'httpd -D FOREGROUND'")
+
+        # The docker image ``githttpd/githttpd`` creates an HTTP server for Git
+        # repositories, using the volume bound onto ``/opt/git-server`` as base
+        # for any subdirectory, the Git URL becoming ``git/{subdirectory-name}}``.
+        yield container, git_repos_path, f'http://localhost:{http_port}/git'
+
+

--- a/test/integration/git/__init__.py
+++ b/test/integration/git/__init__.py
@@ -1,0 +1,18 @@
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Unit tests for Autosubmit Git functions."""

--- a/test/integration/git/test_autosubmit_git.py
+++ b/test/integration/git/test_autosubmit_git.py
@@ -1,0 +1,445 @@
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Integration tests for ``autosubmit_git``."""
+
+from contextlib import nullcontext as does_not_raise
+from getpass import getuser
+from pathlib import Path
+from typing import Callable, ContextManager, Generator, Tuple, Union, cast
+
+import pytest
+from testcontainers.sftp import DockerContainer
+
+from autosubmit.git.autosubmit_git import check_unpushed_changes, clean_git
+from log.log import AutosubmitCritical
+from test.integration.test_utils.git import (
+    create_git_repository, git_commit_all_in_dir, git_clone_repository, git_add_submodule
+)
+
+_EXPID = 'a000'
+
+
+# git operational check
+
+def _get_experiment_data(tmp_path) -> dict:
+    _user = getuser()
+
+    return {
+        'PLATFORMS': {
+            'pytest-ps': {
+                'type': 'ps',
+                'host': '127.0.0.1',
+                'user': _user,
+                'project': 'whatever',
+                'scratch': str(tmp_path / 'scratch'),
+                'DISABLE_RECOVERY_THREADS': 'True'
+            }
+        },
+        'JOBS': {
+            'debug': {
+                'SCRIPT': 'echo "Hello world"',
+                'RUNNING': 'once'
+            },
+        },
+        'PROJECT': {
+            'PROJECT_DESTINATION': 'the_project',
+        },
+        'GIT': {
+            'PROJECT_BRANCH': 'master',
+            'PROJECT_COMMIT': '',
+            'PROJECT_SUBMODULES': False,
+            'FETCH_SINGLE_BRANCH': True
+        }
+    }
+
+
+@pytest.mark.parametrize(
+    "project_type,commit,push,expid,expected",
+    [
+        ('git', False, False, 'o001', pytest.raises(AutosubmitCritical)),
+        ('git', True, False, 'o001', pytest.raises(AutosubmitCritical)),
+        ('git', True, True, 'o001', does_not_raise()),
+        ('git', False, False, 'a001', does_not_raise()),
+        ('git', True, False, 'a001', does_not_raise()),
+        ('git', True, True, 'a001', does_not_raise()),
+    ],
+    ids=[
+        'NOK: Did not commit nor push an operational experiment',
+        'NOK: Committed but did not push an operational experiment',
+        'OK: Commited and pushed an operational experiment',
+        'OK: Did not commit nor push a common experiment',
+        'OK: Committed but did not commit nor push a common experiment',
+        'OK: Committed and pushed a common experiment'
+    ]
+)
+@pytest.mark.docker
+def test_git_local_dirty(
+        project_type: str,
+        commit: bool,
+        push: bool,
+        expid: str,
+        expected: ContextManager,
+        tmp_path: Path,
+        autosubmit_exp: Callable
+) -> None:
+    """Tests that Autosubmit detects dirty local Git repositories, especially with operational experiments."""
+    git_repo = tmp_path / 'git_repository'
+
+    create_git_repository(git_repo, bare=True)
+
+    experiment_data = _get_experiment_data(tmp_path)
+    experiment_data['PROJECT']['PROJECT_TYPE'] = project_type
+    experiment_data['GIT']['PROJECT_ORIGIN'] = f'file://{str(git_repo)}'
+    experiment_data['LOCAL'] = {
+        'PROJECT_PATH': str(git_repo)
+    }
+
+    as_exp = autosubmit_exp(expid, experiment_data)
+    as_conf = as_exp.as_conf
+    proj_dir = Path(as_conf.get_project_dir())
+
+    with open(proj_dir / 'a_file.yaml', 'w') as f:
+        f.write('initial content')
+
+    if project_type == 'git' and commit:
+        git_commit_all_in_dir(proj_dir, push=push)
+
+    with expected:
+        check_unpushed_changes(expid, as_conf)
+
+
+@pytest.mark.parametrize(
+    "commit_git,push_git,commit_submodule,push_submodule,expid,expected",
+    [
+        (False, False, False, False, 'o001', pytest.raises(AutosubmitCritical)),
+        (False, False, True, False, 'o001', pytest.raises(AutosubmitCritical)),
+        (False, False, True, True, 'o001', pytest.raises(AutosubmitCritical)),
+        (True, False, True, True, 'o001', pytest.raises(AutosubmitCritical)),
+        (True, True, True, True, 'o001', does_not_raise()),
+        (False, False, False, False, 'a001', does_not_raise()),
+        (False, False, True, False, 'a001', does_not_raise()),
+        (False, False, True, True, 'a001', does_not_raise()),
+        (True, False, True, True, 'a001', does_not_raise()),
+        (True, True, True, True, 'a001', does_not_raise()),
+    ],
+    ids=[
+        'NOK: Did not commit nor push repository and submodule of an operational experiment',
+        'NOK: Committed the submodule, but did not push it, nor committed or pushed the repository of an operational '
+        'experiment',
+        'NOK: Committed and pushed the submodule, but without committing and pushing the repository of an operational '
+        'experiment',
+        'NOK: Committed and pushed the submodule, committed but did not push the repository of an operational '
+        'experiment',
+        'OK: Committed and pushed the submodule, and committed and pushed the repository of an operational '
+        'experiment',
+        'OK: Did not commit nor push repository and submodule of a common experiment',
+        'OK: Committed the submodule, but did not push it, nor committed or pushed the repository of a common '
+        'experiment',
+        'OK: Committed and pushed the submodule, but without committing and pushing the repository of a common '
+        'experiment',
+        'OK: Committed and pushed the submodule, committed but did not push the repository of a common experiment',
+        'OK: Committed and pushed the submodule, and committed and pushed the repository of a common experiment',
+    ]
+)
+@pytest.mark.docker
+def test_git_submodules_dirty(
+        commit_git: bool,
+        push_git: bool,
+        commit_submodule: bool,
+        push_submodule: bool,
+        expid: str,
+        expected: ContextManager,
+        autosubmit_exp: Callable,
+        git_server: Generator[Tuple[DockerContainer, Path, str], None, None],
+        tmp_path
+) -> None:
+    """Tests that Autosubmit detects dirty local Git submodules, especially with operational experiments.
+
+    This test has a Git repository with a Git submodule. The parameters in this test control whether the
+    Git repository and the Git submodule contents will be committed and pushed.
+
+    If the user has non-committed or non-pushed changes in the repository or submodule, the code is
+    expected to fail, raising an error when the experiment is operational.
+    """
+
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+
+    git_repo = git_repos_path / 'git_repository'
+    git_submodule = git_repos_path / 'git_submodule'
+
+    # Create main repository and submodule repository.
+    create_git_repository(git_repo, bare=True)
+    create_git_repository(git_submodule, bare=True)
+
+    git_repo_url = f'{git_url}/{git_repo.name}'
+    git_submodule_url = f'{git_url}/{git_submodule.name}'
+
+    # Add submodule repository as a submodule in the main repository.
+    temp_clone = tmp_path / 'git_clone'
+    git_clone_repository(git_repo_url, temp_clone)
+    submodule_name = git_submodule.name
+    git_add_submodule(git_submodule_url, temp_clone, name=submodule_name, push=True)
+
+    experiment_data = _get_experiment_data(tmp_path)
+    experiment_data['PROJECT']['PROJECT_TYPE'] = 'git'
+    experiment_data['GIT']['PROJECT_ORIGIN'] = git_repo_url
+    experiment_data['GIT']['PROJECT_SUBMODULES'] = submodule_name
+    experiment_data['LOCAL'] = {
+        'PROJECT_PATH': str(git_repo)
+    }
+
+    as_exp = autosubmit_exp(expid, experiment_data)
+    as_conf = as_exp.as_conf
+    proj_dir = Path(as_conf.get_project_dir())
+
+    with open(proj_dir / submodule_name / 'a_file.yaml', 'w') as f:
+        f.write('modified content')
+
+    if commit_submodule:
+        git_commit_all_in_dir(proj_dir / submodule_name, push=push_submodule)
+
+    if commit_git:
+        git_commit_all_in_dir(proj_dir, push=push_git)
+
+    with expected:
+        check_unpushed_changes(expid, as_conf)
+
+
+@pytest.mark.parametrize(
+    'git_operational_check_enabled,expected',
+    [
+        (True, AutosubmitCritical),
+        (False, 0)
+    ]
+)
+@pytest.mark.docker
+def test_git_operational_experiment_toggle_flag(
+        git_operational_check_enabled: bool,
+        expected: Union[int, Exception],
+        autosubmit_exp: Callable,
+        git_server: Generator[Tuple[DockerContainer, Path, str], None, None],
+        tmp_path,
+        mocker,
+        autosubmit
+) -> None:
+    """Tests running an operational works as expected when the toggle flag is used.
+
+    Users can configure ``CONFIG.GIT_OPERATIONAL_CHECK_ENABLED`` to toggle the Git
+    operational check on and off.
+
+    We launch the same experiment ``o001``, with a dirty Git repository, with and
+    without that feature enabled and confirm that when the Git operational check is
+    enabled, Autosubmit fails to launch the experiment. And when that's off, then
+    Autosubmit ignores that, warning the user about it though.
+    """
+    mocked_log = mocker.patch('autosubmit.autosubmit.Log')
+    expid = 'o001'
+
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+
+    git_repo = git_repos_path / 'git_repository'
+
+    create_git_repository(git_repo, bare=True)
+    git_repo_url = f'{git_url}/{git_repo.name}'
+
+    # Add submodule repository as a submodule in the main repository.
+    temp_clone = tmp_path / 'git_clone'
+    git_clone_repository(git_repo_url, temp_clone)
+
+    experiment_data = _get_experiment_data(tmp_path)
+    experiment_data['PROJECT']['PROJECT_TYPE'] = 'git'
+    experiment_data['GIT']['PROJECT_ORIGIN'] = git_repo_url
+    experiment_data['GIT']['PROJECT_SUBMODULES'] = ''
+
+    if 'CONFIG' not in experiment_data:
+        experiment_data['CONFIG'] = {}
+    experiment_data['CONFIG']['GIT_OPERATIONAL_CHECK_ENABLED'] = git_operational_check_enabled
+
+    as_exp = autosubmit_exp(expid, experiment_data)
+    as_conf = as_exp.as_conf
+    proj_dir = Path(as_conf.get_project_dir())
+
+    # Dirty Git repository.
+    with open(proj_dir / 'a_file.yaml', 'w') as f:
+        f.write('modified content')
+
+    mocker.patch('sys.argv', ['autosubmit', '-lc', 'DEBUG', '-lf', 'DEBUG', 'run', expid])
+    _, args = autosubmit.parse_args()
+
+    if type(expected) is int:
+        exit_code = autosubmit.run_command(args=args)
+        assert exit_code == expected
+
+        assert mocked_log.warning.called
+        assert mocked_log.warning.call_args[0][0] == 'Git operational check disabled by user'
+    else:
+        with pytest.raises(cast(expected, Exception)):
+            autosubmit.run_command(args=args)
+
+
+# -- clean_git
+
+def test_clean_git_not_a_dir(autosubmit_config, mocker):
+    """Test that cleaning Git fails when the project directory is actually a file."""
+    mocked_log = mocker.patch('autosubmit.git.autosubmit_git.Log')
+    as_conf = autosubmit_config(_EXPID, {})
+
+    # Remove proj dir, replacing by a file.
+    proj_dir = Path(as_conf.get_project_dir())
+    proj_dir.parent.mkdir(parents=True, exist_ok=True)
+    proj_dir.touch()
+
+    assert not clean_git(as_conf)
+
+    assert mocked_log.debug.call_args_list[1][0][0] == 'Not a directory... SKIPPING!'
+
+
+def test_clean_git_not_a_git_repo(autosubmit_config, mocker):
+    """Test that cleaning Git fails when the project directory is not a Git repository."""
+    mocked_log = mocker.patch('autosubmit.git.autosubmit_git.Log')
+    as_conf = autosubmit_config(_EXPID, {})
+
+    # Missing .git folder.
+    proj_dir = Path(as_conf.get_project_dir())
+    proj_dir.mkdir(parents=True, exist_ok=True)
+
+    assert not clean_git(as_conf)
+
+    assert mocked_log.debug.call_args_list[1][0][0] == 'Not a git repository... SKIPPING!'
+
+
+def test_clean_git_not_committed(
+        tmp_path,
+        autosubmit_exp,
+        git_server: Generator[Tuple[DockerContainer, Path, str], None, None]
+):
+    """Test that cleaning Git fails when the project directory has new files not committed yet."""
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+
+    git_repo = git_repos_path / test_clean_git_not_committed.__name__
+    create_git_repository(git_repo, bare=True)
+
+    git_repo_url = f'{git_url}/{git_repo.name}'
+
+    experiment_data = _get_experiment_data(tmp_path)
+    experiment_data['PROJECT']['PROJECT_TYPE'] = 'git'
+    experiment_data['GIT']['PROJECT_ORIGIN'] = git_repo_url
+
+    as_exp = autosubmit_exp(_EXPID, experiment_data)
+    as_conf = as_exp.as_conf
+    proj_dir = Path(as_conf.get_project_dir())
+
+    Path(proj_dir / 'not_committed.txt').touch()
+
+    with pytest.raises(AutosubmitCritical) as cm:
+        clean_git(as_conf)
+
+    assert str(cm.value.code) == '7013'
+
+
+def test_clean_git_not_pushed(
+        tmp_path,
+        autosubmit_exp,
+        git_server: Generator[Tuple[DockerContainer, Path, str], None, None]
+):
+    """Test that cleaning Git fails when the project directory has staged changed not pushed."""
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+
+    git_repo = git_repos_path / test_clean_git_not_pushed.__name__
+    create_git_repository(git_repo, bare=True)
+
+    git_repo_url = f'{git_url}/{git_repo.name}'
+
+    experiment_data = _get_experiment_data(tmp_path)
+    experiment_data['PROJECT']['PROJECT_TYPE'] = 'git'
+    experiment_data['GIT']['PROJECT_ORIGIN'] = git_repo_url
+
+    as_exp = autosubmit_exp(_EXPID, experiment_data)
+    as_conf = as_exp.as_conf
+    proj_dir = Path(as_conf.get_project_dir())
+
+    Path(proj_dir / 'not_committed.txt').touch()
+    git_commit_all_in_dir(proj_dir, push=False)
+
+    with pytest.raises(AutosubmitCritical) as cm:
+        clean_git(as_conf)
+
+    assert str(cm.value.code) == '7064'
+
+
+def test_clean_git_set_git_project_commit_fails(
+        tmp_path,
+        autosubmit_exp,
+        mocker,
+        git_server: Generator[Tuple[DockerContainer, Path, str], None, None]
+):
+    """Test that cleaning Git fails when the project commit cannot be recorded."""
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+
+    mocked_log = mocker.patch('autosubmit.git.autosubmit_git.Log')
+
+    git_repo = git_repos_path / test_clean_git_not_pushed.__name__
+    create_git_repository(git_repo, bare=True)
+
+    git_repo_url = f'{git_url}/{git_repo.name}'
+
+    experiment_data = _get_experiment_data(tmp_path)
+    experiment_data['PROJECT']['PROJECT_TYPE'] = 'git'
+    experiment_data['GIT']['PROJECT_ORIGIN'] = git_repo_url
+
+    as_exp = autosubmit_exp(_EXPID, experiment_data)
+    as_conf = as_exp.as_conf
+
+    mocker.patch.object(as_conf, 'set_git_project_commit', return_value=False)
+
+    assert not clean_git(as_conf)
+
+    assert mocked_log.info.call_args_list[1][0][0] == 'Failed to set Git project commit... SKIPPING!'
+
+
+def test_clean_git(
+        tmp_path,
+        autosubmit_exp,
+        mocker,
+        git_server: Generator[Tuple[DockerContainer, Path, str], None, None]
+):
+    """Test that cleaning Git fails when the project commit cannot be recorded."""
+    _, git_repos_path, git_url = git_server  # type: DockerContainer, Path, str
+
+    git_repo = git_repos_path / test_clean_git_not_pushed.__name__
+    create_git_repository(git_repo, bare=True)
+
+    git_repo_url = f'{git_url}/{git_repo.name}'
+
+    experiment_data = _get_experiment_data(tmp_path)
+    experiment_data['PROJECT']['PROJECT_TYPE'] = 'git'
+    experiment_data['GIT']['PROJECT_ORIGIN'] = git_repo_url
+
+    as_exp = autosubmit_exp(_EXPID, experiment_data)
+    as_conf = as_exp.as_conf
+
+    # TODO: remove when bug in config parser is fixed!
+    mocker.patch.object(as_conf, 'set_git_project_commit', return_value=True)
+
+    proj_dir = Path(as_conf.get_project_dir())
+    Path(proj_dir / 'not_committed.txt').touch()
+    git_commit_all_in_dir(proj_dir, push=True)
+
+    assert proj_dir.exists()
+    assert clean_git(as_conf)
+    assert not proj_dir.exists()

--- a/test/integration/scripts/test_clean.py
+++ b/test/integration/scripts/test_clean.py
@@ -166,10 +166,7 @@ def test_clean_git_project(
     mocked_autosubmit_config = mocker.patch('autosubmit.autosubmit.AutosubmitConfig.set_git_project_commit')
     mocked_autosubmit_config.return_value = True
 
-    mocked_autosubmit_git = mocker.patch('autosubmit.autosubmit.AutosubmitGit')
-    mocked_autosubmit_git_obj = mocker.MagicMock()
-    mocked_autosubmit_git.return_value = mocked_autosubmit_git_obj
-    mocked_autosubmit_git_obj.clean_git.return_value = clean_git_return
+    mocker.patch('autosubmit.autosubmit.clean_git', return_value=clean_git_return)
 
     git_project = Path(tmp_path / 'tmp_git_project')
     git_project.mkdir()

--- a/test/integration/test_utils/git.py
+++ b/test/integration/test_utils/git.py
@@ -1,0 +1,137 @@
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Helper functions for working with Git in integration tests."""
+
+from os import system
+from pathlib import Path
+from subprocess import check_output
+from tempfile import TemporaryDirectory
+
+# TODO: Rename branch to main to match the other repositories; note that this
+#       needs to wait until our laptops & CICD env are using Git 2.28+, in order
+#       to use the --initial-branch=main (the -c init.defaultBranch option is
+#       harmless right now, but useless as that's 2.28+ too, thus master for now).
+
+
+def create_git_repository(path: Path, bare=False, branch='master') -> None:
+    """Creates a Git repository.
+
+    Creates the directory if it does not exist.
+
+    :param path: Path to the Git repository.
+    :param bare: Flag to make it a Git bare repository.
+    :param branch: Branch to use.
+    """
+    if not path or path.is_file():
+        raise ValueError(f'You must provide a valid path for your Git repository: {path}')
+
+    if not path.exists():
+        path.mkdir(parents=True)
+
+    git_init_args = []
+    if bare:
+        git_init_args.append('--bare --shared=true')
+
+    # Newer versions of Git have --initial-branch=???, but some of us at the BSC
+    # still have older versions, so we will use a different approach, with
+    # symbolic-ref.
+    commands = [
+        f'git -c init.defaultBranch=master init . {" ".join(git_init_args)}',
+        'git symbolic-ref HEAD refs/heads/master'
+    ]
+
+    command = ';'.join(commands)
+    check_output(command, cwd=str(path), shell=True)
+
+    if bare:
+        # NOTE: You cannot ``git clone file:///location/to/bare/repo -b master``, even
+        #       if the repository exists and has the master branch configured. First,
+        #       you need to push something so Git will create the refs/master for the
+        #       branch. We do it here, since Autosubmit uses ``-b`` by default.
+        with TemporaryDirectory() as td:
+            clone_repo = Path(td, 'clone')
+            git_clone_repository(f'file:///{str(path)}', clone_repo)
+
+            with open(clone_repo / 'README.md', 'w') as f:
+                f.write('This is a test repository of Autosubmit.')
+
+            git_commit_all_in_dir(clone_repo, branch=branch, push=True)
+
+    # git-http-backend (which comes with Git) may have issues with directory
+    # permissions depending on the settings and file system permissions.
+    # Thus, we chmod everything to 0x777 here.
+    system(f'chmod -R 0777 {str(path)}')
+
+
+def git_clone_repository(url: str, path: Path) -> None:
+    """Clone a Git repository into the given path."""
+    if not path or not path.is_absolute() or path.exists():
+        raise ValueError(f'You must provide a valid, absolute, non-existent path for your Git repository: {path}')
+
+    if not url:
+        raise ValueError(f'You must provide a valid URL to be cloned: {url}')
+
+    commands = [
+        f'git clone {url} {str(path)}'
+    ]
+
+    command = ' '.join(commands)
+    # Change cwd even if the path is absolute, just in case...
+    check_output(command, cwd=str(path.parent), shell=True)
+
+
+def git_commit_all_in_dir(path: Path, push=False, remote='origin', branch='master') -> None:
+    """Adds all files in the given path, and creates a single commit for it."""
+    if not path or not path.is_dir():
+        raise ValueError(f'You must provide a valid path for your Git repository: {path}')
+
+    if not list(path.iterdir()):
+        raise ValueError(f'Tried to commit all files, but directory is empty: {path}')
+
+    commands = [
+        'git add .',
+        'git commit -am "Initial commit"'
+    ]
+
+    if push:
+        commands.append(f'git push {remote} {branch}')
+
+    command = ';'.join(commands)
+    check_output(command, cwd=str(path), shell=True)
+
+
+def git_add_submodule(url: str, path: Path, name: str, push=False, remote='origin', branch='master') -> None:
+    """Adds a submodule to a Git repository."""
+    if not path or not path.is_dir():
+        raise ValueError(f'You must provide a valid path for your Git repository: {path}')
+
+    if not url:
+        raise ValueError(f'You must provide a valid URL to be cloned: {url}')
+
+    commands = [
+        'git submodule init',
+        f'git -c protocol.file.allow=always submodule add {url} {name}',
+        'git add .',
+        'git commit -am "Add submodule"'
+    ]
+
+    if push:
+        commands.append(f'git push {remote} {branch}')
+
+    command = ';'.join(commands)
+    check_output(command, cwd=str(path), shell=True)

--- a/test/unit/git/__init__.py
+++ b/test/unit/git/__init__.py
@@ -1,0 +1,18 @@
+# Copyright 2015-2025 Earth Sciences Department, BSC-CNS
+#
+# This file is part of Autosubmit.
+#
+# Autosubmit is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Autosubmit is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Integration tests for Autosubmit Git functions."""

--- a/test/unit/git/test_autosubmit_git.py
+++ b/test/unit/git/test_autosubmit_git.py
@@ -16,13 +16,8 @@
 # along with Autosubmit.  If not, see <http://www.gnu.org/licenses/>.
 
 """Tests for ``AutosubmitGit``."""
-from getpass import getuser
-from pathlib import Path
-from typing import Callable
 
 import pytest
-from pytest_mock import MockerFixture
-from ruamel.yaml import YAML
 
 from autosubmit.git.autosubmit_git import AutosubmitGit
 from log.log import AutosubmitCritical
@@ -198,61 +193,3 @@ def test_copy_code(autosubmit_config, config, mocker, autosubmit):
 )
 def test_valid_git_repo_check(git_repo: str, expected: str) -> None:
     assert AutosubmitGit.is_git_repo(git_repo) == expected
-
-
-def _get_experiment_data(tmp_path):
-    _user = getuser()
-
-    return {
-        'PLATFORMS': {
-            'pytest-ps': {
-                'type': 'ps',
-                'host': '127.0.0.1',
-                'user': _user,
-                'project': 'whatever',
-                'scratch': str(tmp_path / 'scratch'),
-                'DISABLE_RECOVERY_THREADS': 'True'
-            }
-        },
-        'JOBS': {
-            'debug': {
-                'SCRIPT': 'echo "Hello world"',
-                'RUNNING': 'once'
-            },
-        }
-    }
-
-
-@pytest.mark.parametrize(
-    "project_type,expid",
-    [
-        ('git', 'o001'),
-        ('git', 'a001'),
-        ('git', 't001'),
-        ('git', 'e001'),
-    ]
-)
-def test_remote_repo_operational(project_type: str, expid: str, tmp_path: Path, autosubmit_exp: Callable,
-                                 autosubmit, mocker: MockerFixture) -> None:
-    """
-    Tests the check_unpushed_changed function from AutosubmitGit.
-
-    Ensures no operational test with unpushed changes in their Git repository is run.
-    """
-    as_exp = autosubmit_exp(expid, _get_experiment_data(tmp_path))
-    run_dir = Path(as_exp.as_conf.basic_config.LOCAL_ROOT_DIR)
-    temp_path = run_dir / expid / "conf" / f"expdef_{expid}.yml"
-
-    with open(temp_path, 'r') as f:
-        yaml = YAML(typ='rt')
-        data = yaml.load(f)
-    data["PROJECT"]["PROJECT_TYPE"] = project_type
-    with open(temp_path, 'w') as f:
-        yaml.dump(data, f)
-
-    mocker.patch('subprocess.check_output', return_value=b'M\n')
-    if expid[0] != 'o':
-        AutosubmitGit.check_unpushed_changes(expid)
-    else:
-        with pytest.raises(AutosubmitCritical):
-            AutosubmitGit.check_unpushed_changes(expid)

